core/src/main/java/ai/timefold/solver/core/impl/domain/solution/cloner/gizmo/GizmoSolutionClonerImplementor.java

```diff
@@ -269,121 +291,134 @@ protected void createSetSolutionDescriptor(ClassCreator classCreator, SolutionDe
-        BytecodeCreator hasCloneBranch = hasCloneBranchResult.trueBranch();
-        hasCloneBranch.returnValue(maybeClone);
-
-        BytecodeCreator noCloneBranch = hasCloneBranchResult.falseBranch();
-        List<Class<?>> sortedSolutionClassList = new ArrayList<>(solutionClassSet);
-        sortedSolutionClassList.sort(instanceOfComparator);
-
-        BytecodeCreator currentBranch = noCloneBranch;
-        ResultHandle thisObjClass =
-                currentBranch.invokeVirtualMethod(MethodDescriptor.ofMethod(Object.class, "getClass", Class.class), thisObj);
-        for (Class<?> solutionSubclass : sortedSolutionClassList) {
-            ResultHandle solutionSubclassResultHandle = currentBranch.loadClass(solutionSubclass);
-            ResultHandle isSubclass =
-                    currentBranch.invokeVirtualMethod(EQUALS_METHOD, solutionSubclassResultHandle, thisObjClass);
-            BranchResult isSubclassBranchResult = currentBranch.ifTrue(isSubclass);
-
-            BytecodeCreator isSubclassBranch = isSubclassBranchResult.trueBranch();
-
-            GizmoSolutionOrEntityDescriptor solutionSubclassDescriptor =
-                    memoizedSolutionOrEntityDescriptorMap.computeIfAbsent(solutionSubclass,
-                            key -> new GizmoSolutionOrEntityDescriptor(solutionDescriptor, solutionSubclass));
-
-            ResultHandle clone;
-            if (PlanningCloneable.class.isAssignableFrom(solutionSubclass)) {
-                clone = isSubclassBranch.invokeInterfaceMethod(
-                        MethodDescriptor.ofMethod(PlanningCloneable.class, "createNewInstance", Object.class),
-                        thisObj);
-                clone = isSubclassBranch.checkCast(clone, solutionSubclass);
-            } else {
-                clone = isSubclassBranch.newInstance(MethodDescriptor.ofConstructor(solutionSubclass));
-            }
+    private void createCloneSolutionRun(ClonerDescriptor clonerDescriptor,
+            Set<Class<?>> solutionClassSet, Comparator<Class<?>> instanceOfComparator) {
+        var solutionClass = clonerDescriptor.solutionDescriptor.getSolutionClass();
+        var methodCreator =
+                clonerDescriptor.classCreator.getMethodCreator("cloneSolutionRun", solutionClass, solutionClass, Map.class);
+        methodCreator.setModifiers(Modifier.STATIC | Modifier.PUBLIC);
+
+        var thisObj = methodCreator.getMethodParam(0);
+        var solutionNullBranchResult = methodCreator.ifNull(thisObj);
+        try (var solutionIsNullBranch = solutionNullBranchResult.trueBranch()) {
+            solutionIsNullBranch.returnValue(thisObj); // thisObj is null
+        }
 
-            isSubclassBranch.invokeInterfaceMethod(
-                    MethodDescriptor.ofMethod(Map.class, "put", Object.class, Object.class, Object.class),
-                    createdCloneMap, thisObj, clone);
+        try (var solutionIsNotNullBranch = solutionNullBranchResult.falseBranch()) {
```

Comment:
This is a very long branch. I suggest breaking it down into smaller, well-named methods.